src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,1055 @@
+                14: (
+                    gates.M(0, basis=gates.Z),
+                    gates.M(1, basis=gates.Y),
+                ),
+                15: (
+                    gates.M(0, basis=gates.Z),
+                    gates.M(1, basis=gates.Z),
+                ),
+            }
+
+            new_circ = circ.copy()
+
+            for meas in j_to_measurements_2qb[j]:
+                new_circ.add(meas)
+
+            return new_circ
+
+        else:
+            raise ValueError(
+                f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+            )
+
+    else:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+
+def GST_execute_circuit(circuit, k, j, nshots=int(1e4), nqubits=None):
+    """Executes a circuit used in gate set tomography and processes the
+        measurement outcomes for the Pauli Transfer Matrix notation. The circuit
+        should already have noise models implemented, if any, prior to using this
+        function.
+
+        Args:
+        circuit: qibo circuit
+        k (int): The index of the state prepared.
+        j (int): The index of the measurement basis.
+        nshots (int, optional): Number of shots to execute circuit with.
+
+    Returns:
+        numpy.float: Expectation value given by either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
+    """
+
+    if nqubits == 1:
+        # idx_basis = ['I', 'X', 'Y', 'Z']
+        expectation_signs = np.array([[1, 1, 1, 1], [1, -1, -1, -1]])
+        matrix = np.array([[0, 0], [1, 0]])
+
+    elif nqubits == 2:
+        # idx_basis = ['II','IX','IY','IZ', 'XI','XX','XY','XZ', 'YI','YX','YY','YZ', 'ZI','ZX','ZY','ZZ']
+        expectation_signs = np.array(
+            [[1, 1, 1, 1], [1, -1, 1, -1], [1, 1, -1, -1], [1, -1, -1, 1]]
+        )
+        matrix = np.array([[0, 0], [1, 0], [10, 0], [11, 0]])
+
+    else:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+    result = circuit.execute(nshots=nshots)
+    counts = result.frequencies(binary=True)
+
+    # Find value for matrix_jk
+    for v1, v2 in counts.items():
+        row = int(v1, 2)
+        col = v2
+        matrix[row, 1] = col
+
+    probs = matrix[:, 1] / np.sum(matrix[:, 1])
+    matrix = np.hstack((matrix, probs.reshape(-1, 1)))
+
+    k = k + 1
+    j = j + 1
+
+    if j == 1:
+        matrix = np.hstack((matrix, expectation_signs[:, 0].reshape(-1, 1)))
+    elif j == 2 or j == 3 or j == 4:
+        matrix = np.hstack((matrix, expectation_signs[:, 1].reshape(-1, 1)))
+    elif j == 5 or j == 9 or j == 13:
+        matrix = np.hstack((matrix, expectation_signs[:, 2].reshape(-1, 1)))
+    else:
+        matrix = np.hstack((matrix, expectation_signs[:, 3].reshape(-1, 1)))
+
+    temp = matrix[:, 2] * matrix[:, 3]
+    matrix = np.hstack((matrix, temp.reshape(-1, 1)))
+    expectation_val = np.sum(temp)
+
+    return expectation_val
```

Comment:
```suggestion
def GST_execute_circuit(circuit, j, nshots=int(1e4), backend=None):
    """Executes a circuit used in gate set tomography and processes the
        measurement outcomes for the Pauli Transfer Matrix notation. The circuit
        should already have noise models implemented, if any, prior to using this
        function.

        Args:
        circuit: qibo circuit
        k (int): The index of the state prepared.
        j (int): The index of the measurement basis.
        nshots (int, optional): Number of shots to execute circuit with.

    Returns:
        numpy.float: Expectation value given by either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
    """

    if j == 0:
        return 1
    
    nqubits = circuit.nqubits
    if nqubits != 1 and nqubits != 2:
        raise ValueError(
            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
        )

    if backend is None:  # pragma: no cover
        backend = GlobalBackend()

    result = backend.execute_circuit(circuit, nshots=nshots)

    observables=[symbols.I,symbols.Z,symbols.Z,symbols.Z]

    observables_list = list(product(observables,repeat=nqubits))[j]

    observable = 1
    for q, obs in enumerate(observables_list):
        if obs is not symbols.I:
            observable *= obs(q)
    observable = SymbolicHamiltonian(observable,nqubits=nqubits)

    expectation_val = result.expectation_from_samples(observable)

    return expectation_val
```

To calculate the expected values of Pauli strings, it is possible to use `QuantumState.expectation_from_samples(Hamiltonian/SymbolicHamiltonian)`. This method calculates the expected value of diagonal observables in the `Z` basis. To measure all one- or two-qubit Pauli strings, we can diagonalize them beforehand, similar to what you do with `measurement_basis()`, and then compute `<Z>`, which is already diagonal. Therefore, we only need to `call expectation_from_samples()` with observables of the form `Z(0)`, `Z(1)`, or `Z(0)*Z(1)`. It is not necessary to measure `I(0)*I(1)` because, by normalization, it is always 1. Your method of calculating expected values seems correct, but this way, I believe the code is clearer, and we are using methods already implemented in qibo.

Reply:
Thanks. I have tested this version and it works. Let me compile it.